lib/tree.rb

```diff
@@ -14,42 +14,121 @@ class Tree
   attr_reader :root
   def initialize
     @root = nil
+    @nodes = []
+  end
+
+  def add_helper(current, key)
+    if key <= current.key
+      return current if current.left.nil?
+      add_helper(current.left, key)
+    else
+      return current if current.right.nil?
+      add_helper(current.right, key)
+    end
   end
 
   # Time Complexity: 
   # Space Complexity: 
   def add(key, value)
-    raise NotImplementedError
+    if @root == nil
+      @root = TreeNode.new(key, value)
+      return
+    end
+
+    node = add_helper(@root, key)
+    if key <= node.key
+      node.left = TreeNode.new(key, value)
+    else
+      node.right = TreeNode.new(key, value)
+    end
```

Comment:
This doesn't actually work, you only can construct a 3-node tree your way.

Consider:

```suggestion
    @root = add_helper(@root, key, value)
  end
  
  def add_helper(current, key, value)
    return TreeNode.new(key, value) if current.nil?
    
    if key <= current.key
      current.left = add_helper(current.left, key, value)
    else
      current.right = add_helper(current.right, key, value)
    end
    
    return current
  end
```